### PR TITLE
unit_tests: add ringdb unit tests

### PR DESCRIPTION
**Pull request by moneromooo-monero**

### CODE REVIEW

File: tests/unit_tests/ringdb.cpp [hunk 1]
```diff
@@ -0,0 +1,158 @@
...
+  ~RingDB() { boost::filesystem::remove_all(filename); free(filename); }
+
+private:
+  std::string make_filename()
+  {
+    boost::filesystem::path path = tools::get_default_data_dir();
+    path /= "fake";
+    path /= "ringdb-test-XXXXXX";
+    filename = strdup(path.string().c_str());
+    EXPECT_TRUE(mkdtemp(filename)!= NULL);
```
Comment (stoffu) [inline tests/unit_tests/ringdb.cpp]: I'm not familiar with MinGW environment, but `mkdtemp` seems to be unavailable for MinGW-w64 (https://sourceforge.net/p/mingw-w64/code/HEAD/tree/stable/v3.x/mingw-w64-headers/crt/stdlib.h):

```
/monero/tests/unit_tests/ringdb.cpp:83:17: error: 'mkdtemp' was not declared in this scope
     EXPECT_TRUE(mkdtemp(filename)!= NULL);
```

Adding `#include <stdlib.h>` didn't make any difference.

It does seem to exist in the original MinGW (https://sourceforge.net/p/mingw/mingw-org-wsl/ci/mingwrt-3.21.1-release/tree/mingwrt/include/stdlib.h), so this is maybe simply due to the lack of support?

**Comment by stoffu** [comment]
The unit test couldn't run on Mac due to the following error:
```
libc++abi.dylib: terminating with uncaught exception of type boost::exception_detail::clone_impl<boost::exception_detail::error_info_injector<boost::lock_error> >: boost: mutex lock failed in pthread_mutex_lock: Invalid argument
```
The backtrace:
```
* thread #1, queue = 'com.apple.main-thread', stop reason = signal SIGABRT
  * frame #0: 0x00007fff9903ed42 libsystem_kernel.dylib`__pthread_kill + 10
    frame #1: 0x00007fff9912c457 libsystem_pthread.dylib`pthread_kill + 90
    frame #2: 0x00007fff98fa4420 libsystem_c.dylib`abort + 129
    frame #3: 0x00007fff97af894a libc++abi.dylib`abort_message + 266
    frame #4: 0x00007fff97b1dc17 libc++abi.dylib`default_terminate_handler() + 243
    frame #5: 0x00007fff9862c713 libobjc.A.dylib`_objc_terminate() + 124
    frame #6: 0x00007fff97b1ad49 libc++abi.dylib`std::__terminate(void (*)()) + 8
    frame #7: 0x00007fff97b1a7be libc++abi.dylib`__cxa_throw + 121
    frame #8: 0x00000001000b54ab unit_tests`void boost::throw_exception<boost::lock_error>(e=0x00007fff5fbfdb10) at throw_exception.hpp:69
    frame #9: 0x00000001000b55be unit_tests`boost::mutex::lock(this=0x0000000101bdb9f0) at mutex.hpp:119
    frame #10: 0x00000001000b64e3 unit_tests`boost::lock_guard<boost::mutex>::lock_guard(this=0x00007fff5fbfdbc0, m_=0x0000000101bdb9f0) at lock_guard.hpp:38
    frame #11: 0x00000001000b642d unit_tests`boost::lock_guard<boost::mutex>::lock_guard(this=0x00007fff5fbfdbc0, m_=0x0000000101bdb9f0) at lock_guard.hpp:37
    frame #12: 0x00000001004d2cae unit_tests`crypto::rand(N=32, bytes="F\x88\x95\x01\x01") at crypto.h:155
    frame #13: 0x0000000100bccda4 unit_tests`generate_chacha_key() at ringdb.cpp:46
    frame #14: 0x0000000100bd4684 unit_tests`::__cxx_global_var_init.17() at ringdb.cpp:63
    frame #15: 0x0000000100bd655f unit_tests`_GLOBAL__sub_I_ringdb.cpp at ringdb.cpp:0
    frame #16: 0x0000000104442a1b dyld`ImageLoaderMachO::doModInitFunctions(ImageLoader::LinkContext const&) + 385
    frame #17: 0x0000000104442c1e dyld`ImageLoaderMachO::doInitialization(ImageLoader::LinkContext const&) + 40
    frame #18: 0x000000010443e4aa dyld`ImageLoader::recursiveInitialization(ImageLoader::LinkContext const&, unsigned int, char const*, ImageLoader::InitializerTimingList&, ImageLoader::UninitedUpwards&) + 338
    frame #19: 0x000000010443d524 dyld`ImageLoader::processInitializers(ImageLoader::LinkContext const&, unsigned int, ImageLoader::InitializerTimingList&, ImageLoader::UninitedUpwards&) + 138
    frame #20: 0x000000010443d5b9 dyld`ImageLoader::runInitializers(ImageLoader::LinkContext const&, ImageLoader::InitializerTimingList&) + 75
    frame #21: 0x000000010442f47a dyld`dyld::initializeMainExecutable() + 195
    frame #22: 0x00000001044338c6 dyld`dyld::_main(macho_header const*, unsigned long, int, char const**, char const**, char const**, unsigned long*) + 3966
    frame #23: 0x000000010442e249 dyld`dyldbootstrap::start(macho_header const*, int, char const**, long, macho_header const*, unsigned long*) + 470
    frame #24: 0x000000010442e036 dyld`_dyld_start + 54
```

This can be fixed by defining those variables (`KEY_1` etc) as local instead of global.

**Comment by moneromooo-monero** [comment]
Should be fixed now, by making the lock impervious to init order. Please test on mac.

**Comment by stoffu** [comment]
I realized a problem that only shows up when tested on Windows: `RingDB`'s dtor gets called *before* `tools::ringdb`'s dtor, so the call `boost::filesystem::remove_all(filename);` in `RingDB`'s dtor fails because the DB file is still in use (before `mdb_env_close(env);` is called in `tools::ringdb`'s dtor). The following patch fixes this problem, and I've confirmed that the test succeeds on Mac/Windows/Ubuntu.

```diff
diff --git a/tests/unit_tests/ringdb.cpp b/tests/unit_tests/ringdb.cpp
index 33d540857..61e82c1ae 100644
--- a/tests/unit_tests/ringdb.cpp
+++ b/tests/unit_tests/ringdb.cpp
@@ -67,21 +67,16 @@ static const crypto::key_image KEY_IMAGE_1 = generate_key_image();
 static const crypto::public_key OUTPUT_1 = generate_output();
 static const crypto::public_key OUTPUT_2 = generate_output();
 
-class RingDB: public tools::ringdb
+namespace
 {
-public:
-  RingDB(const char *genesis = ""): tools::ringdb(make_filename(), genesis) { }
-  ~RingDB() { boost::filesystem::remove_all(filename); free(filename); }
-
-private:
-  std::string make_filename()
+  char* make_filename()
   {
+    char *filename;
     boost::filesystem::path path = tools::get_default_data_dir();
     path /= "fake";
 #if defined(__MINGW32__) || defined(__MINGW__)
     filename = tempnam(path.string().c_str(), "ringdb-test-");
     EXPECT_TRUE(filename != NULL);
-    filename = strdup(filename);
 #else
     path /= "ringdb-test-XXXXXX";
     filename = strdup(path.string().c_str());
@@ -89,21 +84,28 @@ private:
 #endif
     return filename;
   }
+}
 
-private:
-  char *filename;
-};
+#define MAKE_RINGDB(varname, genesis) \
+  epee::misc_utils::auto_scope_leave_caller cleanup; \
+  char *filename = make_filename(); \
+  tools::ringdb varname(filename, genesis); \
+  cleanup = epee::misc_utils::create_scope_leave_handler([&]() \
+  { \
+    boost::filesystem::remove_all(filename); \
+    free(filename); \
+  })
 
 TEST(ringdb, not_found)
 {
-  RingDB ringdb;
+  MAKE_RINGDB(ringdb, "");
   std::vector<uint64_t> outs;
   ASSERT_FALSE(ringdb.get_ring(KEY_1, KEY_IMAGE_1, outs));
 }
 
 TEST(ringdb, found)
 {
-  RingDB ringdb;
+  MAKE_RINGDB(ringdb, "");
   std::vector<uint64_t> outs, outs2;
   outs.push_back(43); outs.push_back(7320); outs.push_back(8429);
   ASSERT_TRUE(ringdb.set_ring(KEY_1, KEY_IMAGE_1, outs, false));
@@ -113,7 +115,7 @@ TEST(ringdb, found)
 
 TEST(ringdb, convert)
 {
-  RingDB ringdb;
+  MAKE_RINGDB(ringdb, "");
   std::vector<uint64_t> outs, outs2;
   outs.push_back(43); outs.push_back(7320); outs.push_back(8429);
   ASSERT_TRUE(ringdb.set_ring(KEY_1, KEY_IMAGE_1, outs, true));
@@ -126,7 +128,7 @@ TEST(ringdb, convert)
 
 TEST(ringdb, different_genesis)
 {
-  RingDB ringdb;
+  MAKE_RINGDB(ringdb, "");
   std::vector<uint64_t> outs, outs2;
   outs.push_back(43); outs.push_back(7320); outs.push_back(8429);
   ASSERT_TRUE(ringdb.set_ring(KEY_1, KEY_IMAGE_1, outs, false));
@@ -135,21 +137,21 @@ TEST(ringdb, different_genesis)
 
 TEST(blackball, not_found)
 {
-  RingDB ringdb;
+  MAKE_RINGDB(ringdb, "");
   ASSERT_TRUE(ringdb.blackball(OUTPUT_1));
   ASSERT_FALSE(ringdb.blackballed(OUTPUT_2));
 }
 
 TEST(blackball, found)
 {
-  RingDB ringdb;
+  MAKE_RINGDB(ringdb, "");
   ASSERT_TRUE(ringdb.blackball(OUTPUT_1));
   ASSERT_TRUE(ringdb.blackballed(OUTPUT_1));
 }
 
 TEST(blackball, unblackball)
 {
-  RingDB ringdb;
+  MAKE_RINGDB(ringdb, "");
   ASSERT_TRUE(ringdb.blackball(OUTPUT_1));
   ASSERT_TRUE(ringdb.unblackball(OUTPUT_1));
   ASSERT_FALSE(ringdb.blackballed(OUTPUT_1));
@@ -157,7 +159,7 @@ TEST(blackball, unblackball)
 
 TEST(blackball, clear)
 {
-  RingDB ringdb;
+  MAKE_RINGDB(ringdb, "");
   ASSERT_TRUE(ringdb.blackball(OUTPUT_1));
   ASSERT_TRUE(ringdb.clear_blackballs());
   ASSERT_FALSE(ringdb.blackballed(OUTPUT_1));
```

Note that I've also removed `filename = strdup(filename);` because I think the string returned in `filename = tempnam(path.string().c_str(), "ringdb-test-");` is already allocated and needs freeing.

**Comment by moneromooo-monero** [comment]
The body of ~RingDB will be called after ~tools::ringdb. If not, it seems like a bug in your compiler.

**Comment by stoffu** [comment]
@moneromooo-monero I'm fairly certain that the opposite is true: 

> - ~derived
> - ~member derived
> - ~base
> - ~member base

https://stackoverflow.com/a/28356079

AIUI the Mac and Ubuntu compilers (that I saw running the test successfully) should throw in this case.

**Comment by moneromooo-monero** [comment]
You're right, and I had a brain failure there.

**Comment by moneromooo-monero** [comment]
Fixed.

**Pull request merged**